spring-graphql/src/main/java/org/springframework/graphql/data/method/annotation/support/AuthenticationPrincipalArgumentResolver.java

```diff
@@ -0,0 +1,145 @@
+		this.beanResolver = beanResolver;
+	}
+
+	@Override
+	public boolean supportsParameter(MethodParameter parameter) {
+		return findMethodAnnotation(AuthenticationPrincipal.class, parameter) != null;
+	}
+
+	@Override
+	public Object resolveArgument(MethodParameter parameter, DataFetchingEnvironment environment) throws Exception {
+		return getCurrentAuthentication().map(Authentication::getPrincipal)
+				.map((principal) -> resolvePrincipal(parameter, principal))
+				.transform((argument) -> {
+					Class<?> parameterType = parameter.getParameterType();
+					boolean isParameterPublisher = Publisher.class.isAssignableFrom(parameterType);
+					return isParameterPublisher ? Mono.just(argument) : argument;
+				});
+	}
+
+	private Mono<Authentication> getCurrentAuthentication() {
+		SecurityContext securityContext = SecurityContextHolder.getContext();
+		return Mono.justOrEmpty(securityContext.getAuthentication())
+				.switchIfEmpty(ReactiveSecurityContextHolder.getContext().map(SecurityContext::getAuthentication));
+	}
+
+	private Object resolvePrincipal(MethodParameter parameter, Object principal) {
+		AuthenticationPrincipal annotation = findMethodAnnotation(AuthenticationPrincipal.class, parameter);
+		String expressionToParse = annotation.expression();
+		if (StringUtils.hasLength(expressionToParse)) {
+			StandardEvaluationContext context = new StandardEvaluationContext();
+			context.setRootObject(principal);
+			context.setVariable("this", principal);
+			context.setBeanResolver(this.beanResolver);
+			Expression expression = this.parser.parseExpression(expressionToParse);
+			principal = expression.getValue(context);
+		}
+		if (isInvalidType(parameter, principal)) {
+			if (annotation.errorOnInvalidType()) {
+				throw new ClassCastException(principal + " is not assignable to " + parameter.getParameterType());
+			}
+			return null;
+		}
+		return principal;
+	}
+
+	private boolean isInvalidType(MethodParameter parameter, Object principal) {
+		if (principal == null) {
+			return false;
+		}
+		Class<?> typeToCheck = parameter.getParameterType();
+		boolean isParameterPublisher = Publisher.class.isAssignableFrom(parameter.getParameterType());
+		if (isParameterPublisher) {
+			ResolvableType resolvableType = ResolvableType.forMethodParameter(parameter);
+			Class<?> genericType = resolvableType.resolveGeneric(0);
+			if (genericType == null) {
+				return false;
+			}
+			typeToCheck = genericType;
+		}
+		return !ClassUtils.isAssignable(typeToCheck, principal.getClass());
+	}
+
+	/**
+	 * Obtains the specified {@link Annotation} on the specified {@link MethodParameter}.
+	 * @param annotationClass the class of the {@link Annotation} to find on the
+	 * {@link MethodParameter}
+	 * @param parameter the {@link MethodParameter} to search for an {@link Annotation}
+	 * @return the {@link Annotation} that was found or null.
+	 */
+	private <T extends Annotation> T findMethodAnnotation(Class<T> annotationClass, MethodParameter parameter) {
```

Comment:
This method can return `null` but is not marked `@Nullable`.

Reply:
This is now done.